library/core/src/option.rs

```diff
@@ -689,6 +689,92 @@ impl<T> Option<T> {
         }
     }
 
+    const fn get_some_offset() -> isize {
+        if core::mem::size_of::<Option<T>>() == core::mem::size_of::<T>() {
+            // niche optimization means the `T` is always stored at the same position as the Option.
+            0
+        } else {
+            assert!(
+                core::mem::size_of::<Option<T>>()
+                    == core::mem::size_of::<Option<core::mem::MaybeUninit<T>>>()
+            );
+            let some_uninit = Some(core::mem::MaybeUninit::<T>::uninit());
+            // SAFETY: This gets the byte offset of the `Some(_)` value following the fact that
+            // niche optimization is not active, and thus Option<T> and Option<MaybeUninit<t>> share
+            // the same layout, which we have asserted above.
+            unsafe {
+                (some_uninit.as_ref().unwrap() as *const core::mem::MaybeUninit<T> as *const u8)
+                    .offset_from(
+                        &some_uninit as *const Option<core::mem::MaybeUninit<T>> as *const u8,
+                    )
+            }
+        }
+    }
+
+    /// Returns a slice of the contained value, if any. If this is `None`, an
+    /// empty slice is returned. This can be useful to have a single type of
+    /// iterator over an `Option` or slice.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// #![feature(option_as_slice)]
+    ///
+    /// assert_eq!(
+    ///     [Some(42).as_slice(), None.as_slice()],
+    ///     [&[42][..], &[][..]],
+    /// );
+    /// ```
+    #[inline]
+    #[must_use]
+    #[unstable(feature = "option_as_slice", issue = "none")]
+    pub fn as_slice(&self) -> &[T] {
+        // SAFETY: This is sound as long as `get_some_offset` returns the
+        // correct offset. Though in the `None` case, the slice may be located
+        // at a pointer pointing into padding, the fact that the slice is
+        // empty, and the padding is at a properly aligned position for a
+        // value of that type makes it sound.
+        unsafe {
+            core::slice::from_raw_parts(
+                (self as *const Option<T> as *const u8).wrapping_offset(Self::get_some_offset())
+                    as *const T,
+                self.is_some() as usize,
+            )
+        }
+    }
+
+    /// Returns a mutable slice of the contained value, if any. If this is
+    /// `None`, an empty slice is returned. This can be useful to have a
+    /// single type of iterator over an `Option` or slice.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// #![feature(option_as_slice)]
+    ///
+    /// assert_eq!(
+    ///     [Some(42).as_slice_mut(), None.as_slice_mut()],
+    ///     [&mut [42][..], &mut [][..]],
+    /// );
+    /// ```
+    #[inline]
+    #[must_use]
+    #[unstable(feature = "option_as_slice", issue = "none")]
+    pub fn as_slice_mut(&mut self) -> &mut [T] {
```

Comment:
naming: It's [`Vec::as_mut_slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_mut_slice), so I'd suggest using that name here too, rather than this order.

(Basically the idea being that `Option<T>` is essentially `ArrayVec<T, 1>`.)